libpkg/pkg_audit.c

```diff
@@ -542,25 +547,70 @@ pkg_audit_parse_vulnxml(struct pkg_audit *audit)
 {
 	XML_Parser parser;
 	struct vulnxml_userdata ud;
+	uid_t uid, euid;
+	pid_t pid;
+	struct passwd pwd, *res;
+	char pwd_buf[512];
 	int ret = EPKG_OK;
 
-	parser = XML_ParserCreate(NULL);
-	XML_SetElementHandler(parser, vulnxml_start_element, vulnxml_end_element);
-	XML_SetCharacterDataHandler(parser, vulnxml_handle_data);
-	XML_SetUserData(parser, &ud);
+	uid = getuid();
+	euid = geteuid();
+
+	if ((pid = fork()) == 0) {
+		if (uid == 0 || euid != uid) {
+			/*
+			 * User is running pkg audit with root or is using sudo.
+			 * Changing user temporarily to user nobody for safety.
+			 */
+
+			ret = getpwnam_r("nobody", &pwd, pwd_buf, 512, &res);
+			if (res == NULL) {
+				warnx("%s\n", "Error occurred while finding pw entry "
+				    "for user nobody.");
+				errno = ret;
+				exit (-1);
+			}
 
-	ud.cur_entry = NULL;
-	ud.audit = audit;
-	ud.range_num = 0;
-	ud.state = VULNXML_PARSE_INIT;
+			if (setgid(res->pw_gid) != 0)
+				warnx("%s\n", "Could not change egid to nobody.");
 
-	if (XML_Parse(parser, audit->map, audit->len, XML_TRUE) == XML_STATUS_ERROR) {
-		pkg_emit_error("vulnxml parsing error: %s",
-				XML_ErrorString(XML_GetErrorCode(parser)));
-		ret = EPKG_FATAL;
-	}
+			if (setuid(res->pw_uid) != 0)
+				warnx("%s\n", "Could not change euid to nobody.");
+		}
+		parser = XML_ParserCreate(NULL);
+		XML_SetElementHandler(parser, vulnxml_start_element,
+		    vulnxml_end_element);
+		XML_SetCharacterDataHandler(parser, vulnxml_handle_data);
+		XML_SetUserData(parser, &ud);
+
+		ud.cur_entry = NULL;
+		ud.audit = audit;
+		ud.range_num = 0;
+		ud.state = VULNXML_PARSE_INIT;
+
+		if (XML_Parse(parser, audit->map, audit->len, XML_TRUE) ==
+		    XML_STATUS_ERROR) {
+			pkg_emit_error("vulnxml parsing error: %s",
+					XML_ErrorString(XML_GetErrorCode(parser)));
+			ret = EPKG_FATAL;
+		}
 
-	XML_ParserFree(parser);
+		XML_ParserFree(parser);
+		exit(ret);
+	} else if (pid > 0) {
+		if (waitpid(pid, &ret, 0) == -1)
```

Comment:
`errno` should be checked for `EINTR`, which means that you should redo the call to `waitpid`. Otherwise, you might get a zombie process.